src/Microsoft.DocAsCode.Common/Path/PathUtility.cs

```diff
@@ -26,7 +26,7 @@ public static class PathUtility
 
         public static bool IsPathCaseInsensitive()
         {
-            return Environment.OSVersion.Platform < PlatformID.Unix;
+            return true;
```

Comment:
What about update to `Environment.OSVersion.Platform < PlatformID.Unix || Environment.OSVersion.Platform == PlatformID.MacOSX`?

Some considerations:
1. DocFX is still using .NET Framework instead of .NET Core, so I think the environment platform should return `MacOSX` instead of `Unix` here.
2. In Unix, I'd like to keep the behavior as it is used to be. If we return `true` here on Unix, users can write broken links with only casing difference in Markdown without getting any warning from DocFX build, but they are indeed broken links when served on Linux server.

Reply:
Your solution will not work on MacOS (just right tested), because MacOS is reported as Unix (PlatformID.Unix), so docfx assumes that file system is case sensitive (running under mono with .NET Core 3.1.404).

Reply:
> 2. In Unix, I'd like to keep the behavior as it is used to be. If we return `true` here on Unix, users can write broken links with only casing difference in Markdown without getting any warning from DocFX build, but they are indeed broken links when served on Linux server.

I tried to build documentation of [docfx_conflict github project](https://github.com/farcasclaudiu/docfx_conflict) on Linux (Debian 10, mono in docker, docfx 2.56.6). The documentation was built successfully (as expected, because Debian has case sensitive file system). But copying output website to Windows failed, due to "same" file names: docfx_conflict.Member.html and docfx_conflict.member.html.

Reply:
> Your solution will not work on MacOS (just right tested), because MacOS is reported as Unix (PlatformID.Unix), so docfx assumes that file system is case sensitive (running under mono with .NET Core 3.1.404).

Renze means docfx v2 is targeting .NET Framework instead of .NET Core, and running docfx is not officially supported.

> I tried to build documentation of docfx_conflict github project on Linux (Debian 10, mono in docker, docfx 2.56.6). The documentation was built successfully (as expected, because Debian has case sensitive file system).

To be more accurate, ext4, xfs and some other file systems generally used by Linux distributions are case sensitive, NTFS, FAT32, APFS (by default) and ZFS/JFS (with specific option) are case insensitive.

> But copying output website to Windows failed, due to "same" file names: docfx_conflict.Member.html and docfx_conflict.member.html.

I don't know what file system you are using, but I am assuming NTFS since you said Windows. NTFS is a case-insensitive filesystem, you can't have `docfx_conflict.Member.html` and `docfx_conflict.member.html` in same directory and Explorer or other utility will report an error. It's completely unrelated to docfx.

I think there's two things in question here:

- Should docfx generate different results across platforms? If not, we should just treat all platform as case insensitive, and report error on this.
- How does docfx tell a file system is case-sensitive or not? We know a system can run on different file systems, such as one partition is FAT32 and another one is ext4, simply using operating system to determine this could result in runtime error.

Reply:
> Your solution will not work on MacOS (just right tested), because MacOS is reported as Unix (PlatformID.Unix), so docfx assumes that file system is case sensitive (running under mono with .NET Core 3.1.404).

@georgius Got your ideas. So here we cannot rely on `Environment.OSVersion.Platform` to determine if it is Linux or MacOS. I think it is not fair for Linux users if we always return `true` here though.

What we need here is to determine case sensitivity reliably, so [what @yufeih mentioned](https://github.com/dotnet/corefx/blob/bffef76f6af208e2042a2f27bc081ee908bb390b/src/Common/src/System/IO/PathInternal.CaseSensitivity.cs#L37) seems to be the only valid option here...

> Renze means docfx v2 is targeting .NET Framework instead of .NET Core, and running docfx on Mono is not officially supported.

@bianyifan To clarify, Mono is the cross-platform solution recommended by DocFX currently, as DocFX does not take the effort to migrate to .NET Core.

Reply:
@superyyrrzz Should that be interpreted as "You can give Mono a try but we don't guarantee everything works fine"?

Reply:
It should work fine unless DocFX stops supporting cross-platform.

Reply:
> To be more accurate, ext4, xfs and some other file systems generally used by Linux distributions are case sensitive, NTFS, FAT32, APFS (by default) and ZFS/JFS (with specific option) are case insensitive.

Correct, Debian is on EXT4, Windows on NTFS, MacOS on APFS (case insesitive).

> I don't know what file system you are using, but I am assuming NTFS since you said Windows. NTFS is a case-insensitive filesystem, you can't have `docfx_conflict.Member.html` and `docfx_conflict.member.html` in same directory and Explorer or other utility will report an error. It's completely unrelated to docfx.

Correct, Windows is on NTFS, I expected the conflict, I just want to confirm that it happens.

> * Should docfx generate different results across platforms? If not, we should just treat all platform as case insensitive, and report error on this.

By my opinion, on all platforms should be output the same.

On [docfx webpage](https://dotnet.github.io/docfx/) is stated that: 

> DocFX can run on Linux, macOS, and Windows. The generated static website can be deployed to any host such as GitHub Pages or Azure Websites with no additional configuration.

In current state, it is not true, it depends on circumstances.

In my case, I'm using docfx only on Windows and output websites are hosted on IIS (technically, can be hosted also on other web servers). But Linux or MacOS users output websites cannot be in all cases hosted on IIS.

Reply:
Okay, sorry about my misunderstanding before (not familiar with v2...) But Mono implementation have some small differences with .NET Framework. As per [`PlatformID` documentation](https://docs.microsoft.com/en-us/dotnet/api/system.platformid?view=netframework-4.8), macOS should be correctly reported as `MacOSX` instead of `Unix` on .NET Core.

Reply:
In mono is PlatformID overwritten, see https://github.com/mono/mono/blob/9829195f68d0ff07304c30edef7874f4dd6ba6d6/mcs/class/corlib/System/Environment.cs#L228

I think, that critical question is: **Should docfx generate different results across platforms?**

Reply:
> In mono is PlatformID overwritten, see https://github.com/mono/mono/blob/9829195f68d0ff07304c30edef7874f4dd6ba6d6/mcs/class/corlib/System/Environment.cs#L228
> 
> I think, that critical question is: **Should docfx generate different results across platforms?**

Hi, I think the docfx results should be consistent across all platforms. I believe it should behave to the lowest denominator (it should generate files for case INSENSITIVE file systems).

Best regards.

Reply:
> I think the docfx results should be consistent across all platforms.

This is something we would hope to achieve, but given [our current v2 strategy](https://github.com/dotnet/docfx/blob/dev/Roadmap.md#v2-strategy) and the complexity of file systems, we are unlikely to make that happen in v2 at least for linux.

Lets instead focus on _"how to get the correct case sensitivity on Mac"_. Obviously `Environment.OSVersion.Platform` does not work on mono, but `RuntimeInformation.IsOSPlatform(OSPlatform.OSX)` should work.

Reply:
I understand that docfx is already working just fine on Windows (which already has a case insensitive file system).
Is it difficult to have the same behavior as well on other platforms (linux, macOS)?
That would make it consistent across all platforms as results + results files (html) will be portable between various servers (linux, windows) where we wish to host the documentation.

Reply:
```suggestion
            if (System.Runtime.InteropServices.RuntimeInformation.IsOSPlatform(System.Runtime.InteropServices.OSPlatform.Linux))
            {
                return false;
            }

            return true;
```

Is it acceptable?

Reply:
docker is still linux.
Will return true for all platforms will force the docfx engine to generate unique similar filenames for naming conflict cases?

Reply:
> docker is still linux.
> Will return true for all platforms will force the docfx engine to generate unique similar filenames for naming conflict cases?

Will return true for all platforms except Linux, doesn't matter of file system and force the docfx engine to generate unique filenames for naming conflict case.

Reply:
> I understand that docfx is already working just fine on Windows (which already has a case insensitive file system).
Is it difficult to have the same behavior as well on other platforms (linux, macOS)?
That would make it consistent across all platforms as results + results files (html) will be portable between various servers (linux, windows) where we wish to host the documentation.

I totally agree with you about this point, but I will be cautious for this behavior change as it will cause many new warnings complaining invalid link on existing documents.

Reply:
Is it that existing documents can be regenerated with --force parameter and therefore force the linking to be corrected?

Reply:
Yes, it is possible, but I believe some Windows users will not be happy to be forced to fix it, as all case-insensitive links work well before. This is a breaking change we do not want to take per [current strategy in docfx roadmap](https://github.com/dotnet/docfx/blob/dev/Roadmap.md#v2-strategy).